File: src/test/java/kitchenpos/products/tobe/domain/ProductPriceTest.java
package kitchenpos.products.tobe.domain;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ProductPriceTest {

    @DisplayName("상품 가격이 0보다 작을 경우 예외가 발생한다.")
    @ParameterizedTest
    @ValueSource(strings = {"-1", "-1000"})
    void createProductPriceWithNegativePrice() {
        // given
        var price = BigDecimal.valueOf(-1);

        // when, then
        assertThatThrownBy(() -> new ProductPrice(price))
                .isInstanceOf(IllegalArgumentException.class);
    }

    @DisplayName("상품 가격을 생성한다.")
    @ParameterizedTest
    @ValueSource(strings = {"0", "1", "1000"})
    void createProductPrice(String price) {
        // given
        var productPrice = new ProductPrice(new BigDecimal(price));

        // then
        assertThat(productPrice.value()).isEqualTo(new BigDecimal(price));
    }
}

File: src/test/java/kitchenpos/products/tobe/domain/ProductPriceTest.kt
package kitchenpos.products.tobe.domain

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import java.math.BigDecimal

private val Int.won: BigDecimal get() = this.toBigDecimal()

class ProductPriceTest : BehaviorSpec({
    given("상품 금액을 생성할 때") {
        `when`("금액이 0보다 작을 때") {
            val negativeValue = (-1).won

            then("예외가 발생한다") {
                shouldThrow<IllegalArgumentException> {
                    ProductPrice(negativeValue)
                }
            }
        }

        `when`("금액이 0일 때") {
            val zeroValue = 0.won

            then("생성된다") {
                val price = ProductPrice(zeroValue)
                price.value shouldBe zeroValue
            }
        }

        `when`("금액이 0보다 클 때") {
            val positiveValue = 1.won

            then("생성된다") {
                val price = ProductPrice(positiveValue)
                price.value shouldBe positiveValue
            }
        }
    }
})
